### PR TITLE
Create a pool and attach machines to it

### PR DESCRIPTION
**Pull request by huwshimi**
## Done
- Set up the websocket channel to receive additional messages.
- Handle creating a pool and attaching machines to it.

Note: Errors and saving state will be handled by https://github.com/canonical-web-and-design/maas-ui/issues/984 and https://github.com/canonical-web-and-design/maas-ui/issues/881.

## QA
- Visit the machine list. Select machines and choose "Set resource pool..." from the take action menu.
- Change the radiobutton to "Create pool".
- Enter a new pool name and submit the form.
- Your selected machines should be attached to the new pool.

## Fixes
Fixes: https://github.com/canonical-web-and-design/maas-ui/issues/928.

### CODE REVIEW

File: ui/src/app/base/sagas/actions.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  resourcepool as resourcePoolActions,
+} from "app/base/actions";
+
+/**
+ * Generate functions that will use the response to create the dispatchable
+ * action to set the pool for each machine.
+ * @param {Array} machines - A list of machine ids.
+ * @returns {Array} The list of action creator functions.
+ */
+export const generateActionCreators = (machines) =>
```
Comment (Caleb-Ellis) [inline ui/src/app/base/sagas/actions.js:14]: Will every one of these chained websocket sagas need its own function generator like this?

File: ui/src/app/base/sagas/websockets.test.js [hunk 2]
```diff
@@ -284,6 +307,20 @@ describe("websocket sagas", () => {
...
+      request_id: 99,
+      result: { id: 808 },
+    };
+    const action = { type: "NEXT_ACTION" };
+    const actionCreator = jest.fn(() => action);
+    return expectSaga(handleNextActions, response)
+      .provide([[call(getNextActions, 99), [actionCreator]]])
+      .call(actionCreator, response.result)
+      .put(action)
+      .run();
```
Comment (Caleb-Ellis) [inline ui/src/app/base/sagas/websockets.test.js:321]: Gah! I forgot how confusing these saga tests are.

File: ui/src/app/base/sagas/actions.js (hunk 1 above)
Comment (huwshimi) [reply ui/src/app/base/sagas/actions.js:14]: Yeah, something like that. It depends on the action that needs to be performed. The generator is outside. This is only external to the saga to make the saga testable.

[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/base/sagas/actions.js:14]: Oops, yep done.

File: ui/src/app/base/sagas/websockets.js [hunk 3]
```diff
@@ -169,6 +179,40 @@ export function* handleBatch({ request_id, result }) {
...
+    }
+  }
+}
+
+/**
+ * Handle dispatching the next actions, if required.
+ *
+ * @param {Object} response - A websocket response.
+ */
+export function* handleNextActions({ request_id, result }) {
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/base/sagas/websockets.js:202]: Done.

File: ui/src/app/base/sagas/websockets.js [hunk 4]
```diff
@@ -292,7 +340,7 @@ export function* sendMessage(socketClient, action) {
 /**
  * Connect to the WebSocket and watch for message.
  */
-export function* setupWebSocket() {
+export function* setupWebSocket(messageHandlers = []) {
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/base/sagas/websockets.js:343]: Done.

**Comment by huwshimi** [comment]
@squidsoup my plan is that the form will receive the error from the redux store for the pool model, I've created an issue for that here: https://github.com/canonical-web-and-design/maas-ui/issues/984. I need to do https://github.com/canonical-web-and-design/maas-ui/issues/881 first so that the form stays open so it can display the errors.

**Pull request merged**